### DISCUSSION TITLE
Moving3D not working

### DISCUSSION THREAD

**Discussion by kmamor** [gama-platform/gama#344] [Q&A]
I have set up a 3D model. See images

![front_view](https://github.com/user-attachments/assets/f93d89ef-c586-45d7-8499-8698ec1ec7a4)

![top_view](https://github.com/user-attachments/assets/324308ae-5d17-41ec-a14f-02655f47d72c)

and used the following code:

```
model aylesbury3

global {
	
	file shape_file_buildings_comm <- shape_file('../includes/aylesbury_buildings_comm.shp');
	file shape_file_buildings_res <- shape_file('../includes/aylesbury_buildings_res.shp');
	file shape_file_roads_all <- shape_file('../includes/aylesbury_roads_2.shp');
	file shape_file_bounds <- shape_file('../includes/aylesbury_environ.shp');
	int nb_drones <- 10;
	int environment_size <- 20;
	geometry shape <- envelope(shape_file_bounds);
	graph road_network;
	
	init {
		create building_res from: shape_file_buildings_res {
				height <- 10 + rnd(90);
				}
			create building_comm from: shape_file_buildings_comm {
				height <- 10 + rnd(90);
				}
		create road from: shape_file_roads_all;
		road_network <- as_edge_graph(road);
		create drone number: nb_drones {
			 location <- {rnd(1,18000), rnd(1,10000), rnd(1,1000)};	
				}  		} 	}

species drone skills: [moving3D] {
	rgb color;
	
	reflex move {
		do wander;
	}
	aspect base {
		draw sphere(environment_size * 10) color: #orange;
			}  }

species building_res {
rgb color <- #black;
int height ; 
aspect base {
	draw shape color: color depth: height;
}		}

species building_comm {
rgb color <-#black;
int height ; 
aspect base {
	draw shape color: color depth: height;
}		}
	
species road {
	rgb color <- #yellow;

	aspect base {
		draw shape color: color;
	} }

experiment Tuto3D type: gui {
	
	output {
		display map type: 3d {
			species road aspect: base;
			species building_res aspect: base;
			species building_comm aspect: base;
			species drone aspect: base;			
		}  	}  }
```

However the drone reflex - move - does nothing. I run the experiment and nothing occurs. 

Does anyone have any suggestions for things I might need to look at? 

Kevin

**Comment by agrignard** [comment]
Hi Kevin, would it be possible for you to share your project also with the shaepfile so we can run it on your side and try to help you? From what I see the syntax your using (moving3D + wander) should work

**Reply by kmamor** [reply]
H Arnaud,

I have attached the project as a zip file. 

[aylesbury_3.zip](https://github.com/user-attachments/files/17213111/aylesbury_3.zip)


I think the problem is that my shape_file_bounds shapefile is not a 3D file. I am from an ABM background (mainly Netlogo) but want to do a drone based project using GAMA. That has meant trying to also work out how to use QGIS - I believe the issue is that I cannot create a suitable 3D shapefile from the same project that I have taken the roads and buildings (in order that the bounds file has the same coordinates). What I envisaged is trying to form a cube in that QGIS project and saving it as a 3D shapefile. The envelope of that cube (sitting over the buildings and roads) ) would then form the bounds of this GAMA project and form an airspace for drone flights . 

However I cannot figure how to create and save a 3D cube into a shapefile using QGIS that also has the same coordinates as the roads and buildings. 

Any thoughts would be welcome. 

Kevin

**Comment by agrignard** [comment]
Hi Kevin, Thank you for sending the file it helps a lot and it actually raised an issue related to GAMA https://github.com/gama-platform/gama/issues/347

From what I see your code "should" work however surprisingly you might be the first person to use the moving3D skills and wander in a georeferenced world (all the example in the library are in a generic cube of 100,100,100)

In your case  the world is initialized with a 2D shape file so when the drone is trying to move the point it is supposed to move is not "in" the bound of the world so it won"t move.

To keep on working on your project and not being frustrated about the drone not moving maybe one solution can be to  make them move in a 2D world and then add them an altitude location<-{location.x,location.y,z};

**Reply by kmamor** [reply]
Thank you - I'll try that.

Kevin

**Reply by kmamor** [reply]
Arnaud,

I'm trying to get the above code to work as you suggest. I've hit a few problems. 

a) If I use :

```
species drone skills: [moving] {
	rgb color;
	
	reflex move {
		do wander;
	
	}
	aspect base {
		draw sphere(environment_size * 100) color: #orange;
			}

}
```
then the drones do not really move - they just sort of oscillate around their current location. If instead I swap 'wander' with 'move' then then they move? 

b) if I try the following:
```
species drone skills: [moving] {
	rgb color;
	
	reflex move {
		do move;
	location <- {location.x, location.y, location.z + rnd(-10,10)};
	}
	aspect base {
		draw sphere(environment_size * 100) color: #orange;
			}

}
```
then they drift to the right of the world, oscillating up and down. I was expecting:

i. the move action would give them an x and y direction , and then the location <- command would given them a random height location. However they are only getting an increase in their starting x location with no impact on their y location. The z location does seem to move up and down randomly.
